### PR TITLE
DOC: Enforce Numpy Docstring Validation for freqstr, nanos, and rule_code methods

### LINKED ISSUE #58536: DOC: Enforce Numpy Docstring Validation | pandas.tseries.offsets.Week to YearEnd
[DOC: Enforce Numpy Docstring Validation (Parent Issue) #58063](https://github.com/pandas-dev/pandas/issues/58063)

### Documentation problem

Pandas has a script for validating docstrings in code_checks.sh. Currently, some methods fail some of these checks.

### pandas.tseries

https://github.com/pandas-dev/pandas/blob/c468028f5c2398c04d355cef7a8b6a3952620de2/ci/code_checks.sh#L1217-L1261

#### The task is:

1. take 1-5 methods

2. run: `scripts/validate_docstrings.py --format=actions <method-name>`

**example command:** `scripts/validate_docstrings.py --format=actions pandas.Categorical.__array__`
**example output:**

```
################################################################################
################################## Validation ##################################
################################################################################

2 Errors found for `pandas.Categorical.__array__`:
    ES01    No extended summary found
    SA01    See Also section not found
```

3. check if validation docstrings passes for those methods, and if it’s necessary fix the docstrings according to whatever error is reported. Note: We've chosen to ignore ES01 errors, these are not required to be fixed.

4. remove those methods from code_checks.sh if all errors are cleared and the docstring is correct, otherwise, remove the specific error that was fixed from the list of errors for that method.

5. commit, push, open pull request

Please don't comment `take` as multiple people can work on this issue. You also don't need to ask for permission to work on this, just comment on which methods are you going to work : )

If you're new contributor, please check the [contributing guide](https://pandas.pydata.org/docs/dev/development/contributing.html)

thanks @datapythonista @jordan-d-murphy for the inspiration for this

### PR DESCRIPTION
**Pull request by Siniade**
Dosctring Validation for freqstr, nanos, and rule_code methods of pandas.tseries.offsets Week, WeekOfMonth, YearBegin, and YearEnd classes.

- [x] fixes #58536
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added/modified dosctrings for freqstr, nanos, and rule_code methods of pandas.tseries.offsets Week, WeekOfMonth, YearBegin, and YearEnd classes in the pandas/_libs/tslibs/offsets.pyx file.
- [x] Deleted the relevant names from code_checks.sh
- [x] Tested doctstring validation for all the above methods.

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -595,13 +595,36 @@ cdef class BaseOffset:
...
+        tseries.offsets.Day.rule_code : Returns a string representing the base frequency of 'D'.
+
+        Examples
+        --------
+        >>> pd.offsets.Hour().rule_code
+        'h'
+
+        >>> pd.offsets.Week(5).rule_code
+        'W'
+        """  # noqa
```
Comment (mroeschke) [inline pandas/_libs/tslibs/offsets.pyx:613]: Can you fix these docstrings so we don't need a noqa? i.e. Put the `Returns as string...` indented on a new line?

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -986,12 +1039,18 @@ cdef class Tick(SingleConstructorOffset):
...
+        See Also
+        --------
+        tseries.offsets.Hour.nanos :
+            Returns an integer of the total number of nanoseconds.
+        tseries.offsets.Day.nanos :
+            Returns an integer of the total number of nanoseconds.
+        tseries.offsets.WeekOfMonth.nanos :
+            Raises a ValueError because the frequency is non-fixed.
+        tseries.offsets.YearBegin.nanos :
+            Raises a ValueError because the frequency is non-fixed.
```
Comment (mroeschke) [inline pandas/_libs/tslibs/offsets.pyx:1053]:
```suggestion
```

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -779,6 +808,30 @@ cdef class BaseOffset:
 
     @property
     def nanos(self):
+        """
+        Returns a ValueError because the frequency is non-fixed.
```
Comment (mroeschke) [inline pandas/_libs/tslibs/offsets.pyx:812]:
```suggestion
        Returns a integer of the total number of nanoseconds for fixed frequencies.
```

File: pandas/_libs/tslibs/offsets.pyx [hunk 4]
```diff
@@ -779,6 +808,30 @@ cdef class BaseOffset:
...
+        ------
+        ValueError
+            If the frequency is non-fixed.
+
+        See Also
+        --------
+        tseries.offsets.WeekOfMonth.nanos :
+            Raises a ValueError because the frequency is non-fixed.
+        tseries.offsets.YearBegin.nanos :
+            Raises a ValueError because the frequency is non-fixed.
```
Comment (mroeschke) [inline pandas/_libs/tslibs/offsets.pyx:824]:
```suggestion
```

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (Siniade) [reply pandas/_libs/tslibs/offsets.pyx]: Thanks for your review! I have made these changes and also updated the other commits. I am getting a test error which I wasn't getting previously and I'm not sure why.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @Siniade

**Pull request merged**